components/raftstore-v2/src/router/response_channel.rs

```diff
@@ -0,0 +1,448 @@
+    }
+    None
+}
+
+impl<'a, Res> Future for WaitEvent<'a, Res> {
+    type Output = bool;
+
+    #[inline]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let event = &self.core.event;
+        let mut e = event.load(Ordering::Relaxed);
+        let set_bit = set_bit_of(self.event);
+        if let Some(b) = check_bit(e, set_bit) {
+            return Poll::Ready(b);
+        }
+        self.core.waker.register(cx.waker());
+        let subscribed_bit = subscribed_bit_of(self.event);
+        loop {
+            match event.compare_exchange_weak(
+                e,
+                e | subscribed_bit,
+                Ordering::AcqRel,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => return Poll::Pending,
+                Err(v) => e = v,
+            };
+            if let Some(b) = check_bit(e, set_bit) {
+                return Poll::Ready(b);
+            }
+        }
+    }
+}
+
+struct WaitResult<'a, Res> {
+    core: &'a EventCore<Res>,
+}
+
+impl<'a, Res> Future for WaitResult<'a, Res> {
+    type Output = Option<Res>;
+
+    #[inline]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let event = &self.core.event;
+        let set_bit = set_bit_of(PAYLOAD_EVENT);
+        let mut e = event.load(Ordering::Relaxed);
+        if check_bit(e, set_bit).is_some() {
+            unsafe {
+                return Poll::Ready((*self.core.res.get()).take());
+            }
+        }
+        let subscribed_bit = subscribed_bit_of(PAYLOAD_EVENT);
+        self.core.waker.register(cx.waker());
+        loop {
+            match event.compare_exchange_weak(
+                e,
+                e | subscribed_bit,
+                Ordering::AcqRel,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => return Poll::Pending,
+                Err(v) => e = v,
+            };
+            if check_bit(e, set_bit).is_some() {
+                unsafe {
+                    return Poll::Ready((*self.core.res.get()).take());
+                }
+            }
+        }
+    }
+}
+
+pub struct WriteSubscriber {
+    core: Arc<EventCore<RaftCmdResponse>>,
+}
+
+impl WriteSubscriber {
+    pub async fn wait_proposed(&mut self) -> bool {
+        WaitEvent {
+            event: WriteChannel::PROPOSED_EVENT,
+            core: &self.core,
+        }
+        .await
+    }
+
+    pub async fn wait_committed(&mut self) -> bool {
+        WaitEvent {
+            event: WriteChannel::COMMITTED_EVENT,
+            core: &self.core,
+        }
+        .await
+    }
+
+    pub async fn result(&mut self) -> Option<RaftCmdResponse> {
+        WaitResult { core: &self.core }.await
+    }
+}
+
+unsafe impl Send for WriteSubscriber {}
+unsafe impl Sync for WriteSubscriber {}
+
+pub struct WriteChannel {
+    core: ManuallyDrop<Arc<EventCore<RaftCmdResponse>>>,
+}
+
+impl WriteChannel {
+    const PROPOSED_EVENT: u64 = 2;
```

Comment:
all the events should be defined in the same place, to avoid duplicated value.

Reply:
Different channel has different event types. `ReadChannel` doesn't use the event of `WriteChannel`.

Reply:
So can the event be defined as 0, 31 or  > 32 here?  By only looking at WriteChannel itself, we have no idea about that.

Reply:
I will add docs describing the expected range. They are different events for different structs, and will never be mixed used together. It's meaningless to put them together.

Reply:
Probably need some validation check in public APIS for event value.

Reply:
Event value is never public.